text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,137 @@
+
+For a real-world example, consider the `Vec` type in the standard library. It has a `len` field that
+is used to store the number of elements present. Setting this field is exposed publicly in the
+`Vec::set_len` method, which has safety requirements:
+
+- `new_len` must be less than or equal to `capacity()`.
+- The elements at `old_len..new_len` must be initialized.
+
+This field is safe to read, but unsafe to mutate or initialize due to the invariants. These
+invariants cannot be expressed in the type system, so they must be enforced manually. Failure to do
+so may result in undefined behavior elsewhere in `Vec`.
+
+By introducing unsafe fields, Rust can improve the situation where a field that is otherwise safe is
+used as a safety invariant.
+
+# Guide-level explanation
+
+Fields may be declared `unsafe`. Unsafe fields may only be initialized or accessed mutably in an
+unsafe context. Reading the value of an unsafe field may occur in either safe or unsafe contexts. An
+unsafe field may be relied upon as a safety invariant in other unsafe code.
+
+Here is an example to illustrate usage:
+
+```rust
+struct Foo {
+    safe_field: u32,
+    /// Safety: Value must be an odd number.
+    unsafe unsafe_field: u32,
+}
+
+// Unsafe field initialization requires an `unsafe` block.
+// Safety: `unsafe_field` is odd.
+let mut foo = unsafe {
+    Foo {
+        safe_field: 0,
+        unsafe_field: 1,
+    }
+};
+
+// Safe field: no unsafe block.
+foo.safe_field = 1;
+
+// Unsafe field with mutation: unsafe block is required.
+// Safety: The value is odd.
+unsafe { foo.unsafe_field = 3; }
+
+// Unsafe field without mutation: no unsafe block.
+println!("{}", foo.unsafe_field);
+```
+
+For a full description of where a mutable access is considered to have occurred (and why), see
+[RFC 3323]. Keep in mind that due to reborrowing, a mutable access of an unsafe field is not
+necessarily explicit.
+
+[RFC 3323]: https://rust-lang.github.io/rfcs/3323-restrictions.html#where-does-a-mutation-occur
+
+```rust
+fn change_unsafe_field(foo: &mut Foo) {
+    // Safety: An odd number plus two remains an odd number.
+    unsafe { foo.unsafe_field += 2; }
+}
+```
+
+# Reference-level explanation
+
+## Syntax
+
+Using the syntax from [the reference for structs][struct syntax], the change needed to support
+unsafe fields is minimal.
+
+[struct syntax]: https://doc.rust-lang.org/stable/reference/items/structs.html#structs
+
+```diff
+StructField :
+   OuterAttribute*
+   Visibility?
++  unsafe?
+   IDENTIFIER : Type
+
+TupleField :
+   OuterAttribute*
+   Visibility?
++  unsafe?
+   Type
+```
+
+## Behavior
+
+An unsafe field may only be mutated or initialized in an unsafe context. Failure to do so is a compile error.
+
+## "Mutable use" in the compiler
+
+The concept of a "mutable use" [already exists][mutating use] within the compiler. This catches all
+situations that are relevant here, including `ptr::addr_of_mut!`, `&mut`, and direct assignment to a
```

Comment:
Consider this in conjunction with the examples present in the RFC.

```rust
fn make_zero(p: *mut u8) {
    unsafe { *p = 0; }
}

let p = ptr::addr_of_mut!(foo.unsafe_field);
make_zero(p);
```

Ignoring thread-safety, which can be easily achieved with locks, no single step appears wrong. `make_zero` does not do anything wrong — assigning zero to an arbitrary `*mut u8` is fine. Passing a pointer to the method is naturally okay. Yet it still results in unsoundness, as `foo.unsafe_field` must be odd.

Reply:
"assigning zero to an arbitrary `*mut u8` is fine" what‽ No it is not fine‽ An arbitrary `*mut u8` could be null, dangling, aliased... `fn make_zero` is unsound.

Reply:
True — I typed that far too quickly and without thinking. Regardless, it's not immediately obvious to me that `otr::addr_of_mut!` should be allowed safely.

Reply:
I am not fully convinced either way, but I fear it would just be confusing to require `unsafe` for an operation that can't lead to unsoundness, especially as `addr_of!(struct.field) as *mut _` would do the same thing with no `unsafe`.

Reply:
> especially as `addr_of!(struct.field) as *mut _` would do the same thing 

While you can do that, it's undefined behavior to actually mutate the resulting mut pointer. I've just confirmed this with miri.

Reply:
It's UB under Stacked Borrows, but `MIRIFLAGS=-Zmiri-tree-borrows` accepts it.

Reply:
That's surprising. I'm not familiar with tree borrows, admittedly.

Reply:
`ptr::from_mut(struct).wrapping_offset(offset_of!(Struct, field))` should achieve the same thing as `addr_of_mut!`, and I doubt you'd want to make `offset_of` unsafe for unsafe fields. And even without `offset_of`, you could use `ptr::from_mut(struct).wrapping_offset(addr_of!(stuct.Field) - addr_of!(struct))`. Both of these should be safe in every borrowing model.